Remove enterprise license CRD

pebrc opened the pull request:
This replaces the Enterprise license CRD with a regular secret. 

* moves parsing from CLI tool into controllers (better usability)
* models Enterprise license as simple struct 
* stores the serialized form of that license struct in a k8s secret
* add helper functions to retrieve and store these license secrets 
* modifies validation webhook to target license secrets instead of license CRDs

Related to #607 


The operator level ECK trial is started by the user by submitting a secret of this form: 

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: any-name-you-like
  namespace: elastic-system
  labels:
    licencse.k8s.elastic.co/type: enterprise-trial
  annotations:
    elastic.co/eula: accepted
```

sebgl commented on operators/pkg/controller/license/trial/trial_controller_integration_test.go:
```diff
@@ -78,45 +67,43 @@ func TestReconcile(t *testing.T) {
+	// require.NoError(t, license.UpdateEnterpriseLicense(c, types.NamespacedName{
+	// 	Namespace: operatorNs,
+	// 	Name:      string(license.LicenseTypeEnterpriseTrial),
+	// }, trialLicense))
```
to be removed?

sebgl commented on operators/pkg/controller/license/validation/validations.go:
```diff
@@ -5,32 +5,26 @@
-		return validation.Result{Allowed: false, Reason: err.Error()}
+
+	if ctx.Proposed.Annotations[license.EULAAnnotation] != license.EULAAcceptedValue {
+		return validation.Result{Allowed: false, Reason: "Please set the annotation elastic.co/eula to accepted to accept the EULA"}
```
Maybe quote "accepted"

thbkrkr replied:
+1 to add quotes.
Maybe it also deserves its constant since it's used three times.

thbkrkr commented on operators/pkg/controller/common/license/check.go:
```diff
@@ -12,58 +12,84 @@ import (
+	Valid(l EnterpriseLicense) (bool, error)
+}
+
 // Checker contains parameters for license checks.
```
godoc: checker in lowercase

thbkrkr commented on operators/pkg/controller/common/license/crud.go:
```diff
@@ -0,0 +1,122 @@
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	util_errors "k8s.io/apimachinery/pkg/util/errors"
```
According to the go doc, the use of underscores in a go package name should be avoided. I think I would go with `k8serrors` here.

thbkrkr commented on operators/pkg/controller/common/license/parser.go:
```diff
@@ -0,0 +1,22 @@
+	var license EnterpriseLicense
+	err := json.Unmarshal(raw[FileName], &license)
+	if err != nil {
+		return EnterpriseLicense{}, errors.Wrapf(err, "License cannot be unmarshalled")
```
Error strings should not be capitalized.

thbkrkr commented on operators/pkg/controller/license/license_controller.go:
```diff
@@ -38,7 +37,8 @@ const (
 	// In case of any operational issues affecting this controller clusters will have enough runway on their current license.
-	defaultSafetyMargin = 30 * 24 * time.Hour
+	defaultSafetyMargin   = 30 * 24 * time.Hour
+	minimumRetryInternval = 1 * time.Hour
```
typo: `s|minimumRetryInternval|minimumRetryInterval|`

thbkrkr commented on operators/pkg/controller/common/license/trial.go:
```diff
@@ -54,49 +51,52 @@ func InitTrial(c k8s.Client, l *estype.EnterpriseLicense) (*rsa.PublicKey, error
 	}
 	err = c.Create(&trialStatus)
 	if err != nil {
 		return nil, pkgerrors.Wrap(err, "Failed to create trial status")
```
Error strings should not be capitalized.

thbkrkr commented on operators/pkg/controller/common/license/trial.go:
```diff
@@ -25,18 +23,17 @@ import (
 	}
 
 	if err := populateTrialLicense(l); err != nil {
 		return nil, pkgerrors.Wrap(err, "Failed to populate trial license")
```
Error strings should not be capitalized.

pebrc commented:
@sebgl or @thbkrkr If you could take another quick look. I fixed an issue with the validations discovered during testing and modified the label handling and license 'detection' slightly to rely on the more specific `license.k8s.elastic.co/type` instead of the more general `common.k8s.elastic.co/type` which simplifies things and makes it easier for users to create the correct secret
